File: firebase-database/src/main/java/com/androidhuman/rxfirebase3/database/RxFirebaseDatabase.java
package com.androidhuman.rxfirebase3.database;

import com.google.firebase.database.DataSnapshot;
import com.google.firebase.database.DatabaseReference;
import com.google.firebase.database.GenericTypeIndicator;
import com.google.firebase.database.MutableData;
import com.google.firebase.database.Query;
import com.google.firebase.database.Transaction;

import com.androidhuman.rxfirebase3.database.model.DataValue;
import com.androidhuman.rxfirebase3.database.transformers.SingleTransformerOfClazz;
import com.androidhuman.rxfirebase3.database.transformers.SingleTransformerOfGenericTypeIndicator;
import com.androidhuman.rxfirebase3.database.transformers.TransformerOfClazz;
import com.androidhuman.rxfirebase3.database.transformers.TransformerOfGenericTypeIndicator;

import android.support.annotation.CheckResult;
import android.support.annotation.NonNull;

import java.util.Map;

import io.reactivex.rxjava3.core.BackpressureStrategy;
import io.reactivex.rxjava3.core.Completable;
import io.reactivex.rxjava3.core.Flowable;
import io.reactivex.rxjava3.core.Observable;
import io.reactivex.rxjava3.core.Single;
import io.reactivex.rxjava3.functions.Function;

public final class RxFirebaseDatabase {

    @NonNull
    @CheckResult
    public static Observable<ChildEvent> childEvents(@NonNull DatabaseReference ref) {
        return new ChildEventsObserver(ref);
    }

    @NonNull
    @CheckResult
    public static Flowable<ChildEvent> childEvents(
            @NonNull DatabaseReference ref, BackpressureStrategy strategy) {
        return childEvents(ref).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static Observable<ChildEvent> childEvents(@NonNull Query query) {
        return new QueryChildEventsObserver(query);
    }

    @NonNull
    @CheckResult
    public static Flowable<ChildEvent> childEvents(
            @NonNull Query query, BackpressureStrategy strategy) {
        return childEvents(query).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static Single<DataSnapshot> data(@NonNull DatabaseReference ref) {
        return new DataObserver(ref);
    }

    @NonNull
    @CheckResult
    public static Single<DataSnapshot> data(@NonNull Query query) {
        return new QueryObserver(query);
    }

    @NonNull
    @CheckResult
    public static Observable<DataSnapshot> dataChanges(@NonNull DatabaseReference ref) {
        return new DataChangesObserver(ref);
    }

    @NonNull
    @CheckResult
    public static Flowable<DataSnapshot> dataChanges(
            @NonNull DatabaseReference ref, BackpressureStrategy strategy) {
        return dataChanges(ref).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static Observable<DataSnapshot> dataChanges(@NonNull Query query) {
        return new QueryChangesObserver(query);
    }

    @NonNull
    @CheckResult
    public static Flowable<DataSnapshot> dataChanges(
            @NonNull Query query, BackpressureStrategy strategy) {
        return dataChanges(query).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull DatabaseReference ref, @NonNull Class<T> clazz) {
        return dataChanges(ref).compose(new TransformerOfClazz<>(clazz));
    }

    @NonNull
    @CheckResult
    public static <T> Flowable<DataValue<T>> dataChangesOf(
            @NonNull DatabaseReference ref, @NonNull Class<T> clazz,
            BackpressureStrategy strategy) {
        return dataChangesOf(ref, clazz).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull Query query, @NonNull Class<T> clazz) {
        return dataChanges(query).compose(new TransformerOfClazz<>(clazz));
    }

    @NonNull
    @CheckResult
    public static <T> Flowable<DataValue<T>> dataChangesOf(
            @NonNull Query query, @NonNull Class<T> clazz,
            BackpressureStrategy strategy) {
        return dataChangesOf(query, clazz).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull DatabaseReference ref, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return dataChanges(ref)
                .compose(new TransformerOfGenericTypeIndicator<>(typeIndicator));
    }

    @NonNull
    @CheckResult
    public static <T> Flowable<DataValue<T>> dataChangesOf(
            @NonNull DatabaseReference ref, @NonNull GenericTypeIndicator<T> typeIndicator,
            BackpressureStrategy strategy) {
        return dataChangesOf(ref, typeIndicator).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static <T> Observable<DataValue<T>> dataChangesOf(
            @NonNull Query query, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return dataChanges(query)
                .compose(new TransformerOfGenericTypeIndicator<>(typeIndicator));
    }

    @NonNull
    @CheckResult
    public static <T> Flowable<DataValue<T>> dataChangesOf(
            @NonNull Query query, @NonNull GenericTypeIndicator<T> typeIndicator,
            BackpressureStrategy strategy) {
        return dataChangesOf(query, typeIndicator).toFlowable(strategy);
    }

    @NonNull
    @CheckResult
    public static <T> Single<T> dataOf(
            @NonNull DatabaseReference ref, @NonNull Class<T> clazz) {
        return data(ref).compose(new SingleTransformerOfClazz<>(clazz));
    }

    @NonNull
    @CheckResult
    public static <T> Single<T> dataOf(
            @NonNull Query query, @NonNull Class<T> clazz) {
        return data(query).compose(new SingleTransformerOfClazz<>(clazz));
    }

    @NonNull
    @CheckResult
    public static <T> Single<T> dataOf(
            @NonNull DatabaseReference ref, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return data(ref).compose(new SingleTransformerOfGenericTypeIndicator<>(typeIndicator));
    }

    @NonNull
    @CheckResult
    public static <T> Single<T> dataOf(
            @NonNull Query query, @NonNull GenericTypeIndicator<T> typeIndicator) {
        return data(query).compose(new SingleTransformerOfGenericTypeIndicator<>(typeIndicator));
    }

    @NonNull
    @CheckResult
    public static Completable removeValue(@NonNull DatabaseReference ref) {
        return new RemoveValueObserver(ref);
    }

    @NonNull
    @CheckResult
    public static <T> Completable setPriority(
            @NonNull DatabaseReference ref, @NonNull T priority) {
        return new SetPriorityObserver<>(ref, priority);
    }

    @NonNull
    @CheckResult
    public static <T> Completable setValue(
            @NonNull DatabaseReference ref, @NonNull T value) {
        return new SetValueObserver<>(ref, value);
    }

    @NonNull
    @CheckResult
    public static <T, U> Completable setValue(
            @NonNull DatabaseReference ref, @NonNull T value, @NonNull U priority) {
        return new SetValueWithPriorityObserver<>(ref, value, priority);
    }

    @NonNull
    @CheckResult
    public static Completable runTransaction(
            @NonNull DatabaseReference ref,
            @NonNull Function<MutableData, Transaction.Result> task) {
        return runTransaction(ref, true, task);
    }

    @NonNull
    @CheckResult
    public static Completable runTransaction(
            @NonNull DatabaseReference ref, boolean fireLocalEvents,
            @NonNull Function<MutableData, Transaction.Result> task) {
        return new RunTransactionObserver(ref, fireLocalEvents, task);
    }

    @NonNull
    @CheckResult
    public static Completable updateChildren(
            @NonNull DatabaseReference ref, @NonNull Map<String, Object> update) {
        return new UpdateChildrenObserver(ref, update);
    }

    private RxFirebaseDatabase() {
        throw new AssertionError("No instances");
    }
}

File: firebase-database-kotlin/src/main/kotlin/com/androidhuman/rxfirebase2/database/RxFirebaseDatabase.kt
@file:Suppress("NOTHING_TO_INLINE", "UNUSED")

package com.androidhuman.rxfirebase2.database

import com.androidhuman.rxfirebase2.database.model.DataValue
import com.google.firebase.database.DataSnapshot
import com.google.firebase.database.DatabaseReference
import com.google.firebase.database.GenericTypeIndicator
import com.google.firebase.database.MutableData
import com.google.firebase.database.Query
import com.google.firebase.database.Transaction
import io.reactivex.BackpressureStrategy
import io.reactivex.Completable
import io.reactivex.Flowable
import io.reactivex.Observable
import io.reactivex.Single

inline fun DatabaseReference.childEvents()
        : Observable<ChildEvent>
        = RxFirebaseDatabase.childEvents(this)

inline fun DatabaseReference.chileEvents(strategy: BackpressureStrategy)
        : Flowable<ChildEvent>
        = RxFirebaseDatabase.childEvents(this, strategy)

inline fun DatabaseReference.data()
        : Single<DataSnapshot>
        = RxFirebaseDatabase.data(this)

inline fun DatabaseReference.dataChanges()
        : Observable<DataSnapshot>
        = RxFirebaseDatabase.dataChanges(this)

inline fun DatabaseReference.dataChanges(strategy: BackpressureStrategy)
        : Flowable<DataSnapshot>
        = RxFirebaseDatabase.dataChanges(this, strategy)

inline fun <reified T : Any> DatabaseReference.dataChangesOf()
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, T::class.java)

inline fun <reified T : Any> DatabaseReference.dataChangesOf(strategy: BackpressureStrategy)
        : Flowable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, T::class.java, strategy)

inline fun <reified T : Any> DatabaseReference.dataChangesOf(typeIndicator: GenericTypeIndicator<T>)
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, typeIndicator)

inline fun <reified T : Any> DatabaseReference.dataChangesOf(
        typeIndicator: GenericTypeIndicator<T>, strategy: BackpressureStrategy)
        : Flowable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, typeIndicator, strategy)

inline fun <reified T : Any> DatabaseReference.dataOf()
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, T::class.java)

inline fun <reified T : Any> DatabaseReference.dataOf(typeIndicator: GenericTypeIndicator<T>)
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, typeIndicator)

inline fun DatabaseReference.rxRemoveValue()
        : Completable
        = RxFirebaseDatabase.removeValue(this)

inline fun DatabaseReference.rxRunTransaction(noinline task: (MutableData) -> Transaction.Result)
        : Completable
        = RxFirebaseDatabase.runTransaction(this, task)

inline fun DatabaseReference.rxRunTransaction(
        fireLocalEvents: Boolean, noinline task: (MutableData) -> Transaction.Result)
        : Completable
        = RxFirebaseDatabase.runTransaction(this, fireLocalEvents, task)

inline fun <reified T : Any> DatabaseReference.rxSetPriority(priority: T)
        : Completable
        = RxFirebaseDatabase.setPriority(this, priority)

inline fun <reified T : Any> DatabaseReference.rxSetValue(value: T)
        : Completable
        = RxFirebaseDatabase.setValue(this, value)

inline fun <reified T : Any, reified U : Any> DatabaseReference.rxSetValue(value: T, priority: U)
        : Completable
        = RxFirebaseDatabase.setValue(this, value, priority)

inline fun DatabaseReference.rxUpdateChildren(update: Map<String, Any?>)
        : Completable
        = RxFirebaseDatabase.updateChildren(this, update)

inline fun Query.childEvents()
        : Observable<ChildEvent>
        = RxFirebaseDatabase.childEvents(this)

inline fun Query.childEvents(strategy: BackpressureStrategy)
        : Flowable<ChildEvent>
        = RxFirebaseDatabase.childEvents(this, strategy)

inline fun Query.data()
        : Single<DataSnapshot>
        = RxFirebaseDatabase.data(this)

inline fun Query.dataChanges()
        : Observable<DataSnapshot>
        = RxFirebaseDatabase.dataChanges(this)

inline fun Query.dataChanges(strategy: BackpressureStrategy)
        : Flowable<DataSnapshot>
        = RxFirebaseDatabase.dataChanges(this, strategy)

inline fun <reified T : Any> Query.dataChangesOf()
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, T::class.java)

inline fun <reified T : Any> Query.dataChangesOf(strategy: BackpressureStrategy)
        : Flowable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, T::class.java, strategy)

inline fun <reified T : Any> Query.dataChangesOf(typeIndicator: GenericTypeIndicator<T>)
        : Observable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, typeIndicator)

inline fun <reified T : Any> Query.dataChangesOf(
        typeIndicator: GenericTypeIndicator<T>, strategy: BackpressureStrategy)
        : Flowable<DataValue<T>>
        = RxFirebaseDatabase.dataChangesOf(this, typeIndicator, strategy)

inline fun <reified T : Any> Query.dataOf()
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, T::class.java)

inline fun <reified T : Any> Query.dataOf(typeIndicator: GenericTypeIndicator<T>)
        : Single<T>
        = RxFirebaseDatabase.dataOf(this, typeIndicator)
